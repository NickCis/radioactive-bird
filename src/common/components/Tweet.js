import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from 'material-ui/styles';
import Card, { CardHeader, CardMedia, CardContent, CardActions } from 'material-ui/Card';
import Avatar from 'material-ui/Avatar';
import IconButton from 'material-ui/IconButton';
import Typography from 'material-ui/Typography';
import red from 'material-ui/colors/red';
import FavoriteIcon from 'material-ui-icons/Favorite';
import ShareIcon from 'material-ui-icons/Share';
import Link from 'material-ui-icons/Link';

const styles = theme => ({
  card: {
    maxWidth: 760,
  },
  media: {
    height: 194,
  },
  expand: {
    transform: 'rotate(0deg)',
    transition: theme.transitions.create('transform', {
      duration: theme.transitions.duration.shortest,
    }),
  },
  avatar: {
    backgroundColor: red[500],
  },
  flexGrow: {
    flex: '1 1 auto',
  },
  actionButton: {
    'pointer-events': 'none',
  },
});

class Tweet extends React.Component {
  getCardHeaderTitle() {
    const { name, screen_name } = this.props.tweet.user;
    return `${name} @${screen_name}`;
  }

  getCardHeaderSubHeader() {
    const { created_at } = this.props.tweet;
    const date = new Date(created_at);
    return date.toLocaleDateString('en-US', {
      year: 'numeric',
      month: 'long',
      day: 'numeric',
    });
  }

  getCardHeaderAvatar() {
    const { classes, tweet: { user: { name, profile_image_url, screen_name } } } = this.props;
    if (profile_image_url) {
      const url = profile_image_url.replace(/^https?:/, '');
      return (
        <Avatar alt={name} src={url} />
      );
    }

    return (
      <Avatar alt={name} className={classes.avatar}>
        { screen_name[0] }
      </Avatar>
    );
  }

  getCardMediaPhoto(media, i) {
    const { classes, tweet: { text } } = this.props;
    const url = media.media_url.replace(/^https?:/, '');
    return (
        <CardMedia
          key={i}
          className={classes.media}
          image={url}
          title={text}
        />
    );
  }

  getCardMedia() {
    const { media } = this.props.tweet.entities || {};
    if (! media)
      return null;

    return media.map((m, i) => {
      switch (m.type) {
        case 'photo':
          return this.getCardMediaPhoto(m, i);

        default:
          console.log(`Unknown media type: ${m.type}`);
          return null;
      }

    });
  }

  render() {
    const { classes, className, tweet: { favorite_count, retweet_count, text } } = this.props;

    return (
      <Card className={[classes.card, className ? className : ''].join(' ')}>
        <CardHeader
          avatar={this.getCardHeaderAvatar()}
          title={this.getCardHeaderTitle()}
          subheader={this.getCardHeaderSubHeader()}
        />
        { this.getCardMedia() }
        <CardContent>
          <Typography component="p">
            { text }
          </Typography>
        </CardContent>
        <CardActions disableActionSpacing>
          <IconButton className={classes.actionButton}>
            <FavoriteIcon />
            <Typography type="caption">
              { favorite_count }
            </Typography>
          </IconButton>
          <IconButton className={classes.actionButton}>
            <ShareIcon />
            <Typography type="caption">
              { retweet_count }
            </Typography>
          </IconButton>
          <div className={classes.flexGrow} />
          <IconButton aria-label="Go to Tweet">
            <Link />
          </IconButton>
        </CardActions>
      </Card>
    );
  }
}

Tweet.propTypes = {
  tweet: PropTypes.shape({
    user: PropTypes.shape({
      name: PropTypes.string.isRequired,
      screen_name: PropTypes.string.isRequired,
      profile_image_url: PropTypes.string,
    }),
    favorite_count: PropTypes.number.isRequired,
    retweet_count: PropTypes.number.isRequired,
    text: PropTypes.string.isRequired,
    created_at: PropTypes.string.isRequired,
    entities: PropTypes.shape({
      media: PropTypes.array,
    }),
  }),
  classes: PropTypes.object.isRequired,
  className: PropTypes.string,
};

export default withStyles(styles)(Tweet);
