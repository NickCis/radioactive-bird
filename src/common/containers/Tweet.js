import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from 'material-ui/styles';
import { CircularProgress } from 'material-ui/Progress';
import TweetComponent from '../components/Tweet';
import connectWithSSR from '../components/connectWithSSR';
import { bindActionCreators } from 'redux';
import { fetchTweetIfNeeded, fetchTweet } from '../actions/tweets';

const styles = {
  wrapper: {
    margin: '15px',
  },
  tweet: {
    margin: '0 auto',
  },
  loading: {
    'margin-top': '15px',
    width: '100%',
    'text-align': 'center',
  },
};

export class Tweet extends React.Component {
  static getInitialData({fetchTweetIfNeeded, match}) {
    if (!match || !match.params || !match.params.id)
      return Promise.resolve();

    return fetchTweetIfNeeded(match.params.id);
  }

  static get propTypes() {
    return {
      match: PropTypes.object.isRequired,
      tweets: PropTypes.object.isRequired,
      fetchTweetIfNeeded: PropTypes.func.isRequired,
    };
  }

  isLoading() {
    const tweet = this.getTweet();
    if (! tweet)
      return true;

    return tweet.loading;
  }

  getId() {
    return this.props.match.params.id;
  }

  getTweet() {
    return this.props.tweets[this.getId()];
  }

  loading() {
    const { classes } = this.props;
    return (
      <div className={classes.loading}>
        <CircularProgress />
      </div>
    );
  }

  render() {
    const { classes } = this.props;
    if (this.isLoading())
      return this.renderLoading();

    return (
      <div className={classes.wrapper}>
        <TweetComponent className={classes.tweet} tweet={this.getTweet()} hideLink={true}/>
      </div>
    );
  }
}

export const StyledTweet = withStyles(styles)(Tweet);

const mapStateToProps = state => ({
  tweets: state.tweets,
});

const mapDispatchToProps = dispatch => bindActionCreators({
  fetchTweetIfNeeded,
  fetchTweet,
}, dispatch);

export default connectWithSSR(mapStateToProps, mapDispatchToProps)(StyledTweet);
