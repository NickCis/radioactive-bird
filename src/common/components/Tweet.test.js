import React from 'react';
import { createRender } from 'material-ui/test-utils';
import Tweet from './Tweet';

describe('<Tweet />', () => {
  const render = createRender();

  it('renders Tweet correctly', () => {
    const wrapper = render(
      <Tweet
        tweet={{
          title: 'test',
          user: {
            screen_name: 'test',
            name: 'Test',
          },
          favorite_count: 0,
          retweet_count: 0,
          text: 'test',
          created_at: 'Sun Sep 03 20:05:30 +0000 2017',
        }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
