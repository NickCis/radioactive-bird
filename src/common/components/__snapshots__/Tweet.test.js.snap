// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tweet /> renders Tweet correctly 1`] = `
<div
  class="MuiPaper-root-8 MuiPaper-shadow2-12 MuiPaper-rounded-9 Tweet-card-1 "
>
  <div
    class="MuiCardContent-root-40 MuiCardHeader-root-35"
  >
    <div
      class="MuiCardHeader-avatar-36"
    >
      <div
        class="MuiAvatar-root-41 MuiAvatar-colorDefault-42 Tweet-avatar-4"
      >
        t
      </div>
    </div>
    <div
      class="MuiCardHeader-content-37"
    >
      <span
        class="MuiTypography-root-44 MuiTypography-body2-52 MuiCardHeader-title-38"
      >
        Test @test
      </span>
      <span
        class="MuiTypography-root-44 MuiTypography-body2-52 MuiTypography-colorSecondary-65 MuiCardHeader-subheader-39"
      >
        September 3, 2017
      </span>
    </div>
  </div>
  <div
    class="MuiCardContent-root-40"
  >
    <p
      class="MuiTypography-root-44 MuiTypography-body1-53"
    >
      test
    </p>
  </div>
  <div
    class="MuiCardActions-root-68"
  >
    <button
      class="MuiButtonBase-root-79 MuiIconButton-root-70 Tweet-actionButton-6"
      role="button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label-76"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-81 MuiIconButton-icon-77"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
          >
            
          </path>
        </svg>
        <span
          class="MuiTypography-root-44 MuiTypography-caption-54"
        >
          0
        </span>
      </span>
      <span
        class="MuiTouchRipple-root-82"
      >
        
      </span>
    </button>
    <button
      class="MuiButtonBase-root-79 MuiIconButton-root-70 Tweet-actionButton-6"
      role="button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label-76"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-81 MuiIconButton-icon-77"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81 1.66 0 3-1.34 3-3s-1.34-3-3-3-3 1.34-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9c-1.66 0-3 1.34-3 3s1.34 3 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.16c-.05.21-.08.43-.08.65 0 1.61 1.31 2.92 2.92 2.92 1.61 0 2.92-1.31 2.92-2.92s-1.31-2.92-2.92-2.92z"
          >
            
          </path>
        </svg>
        <span
          class="MuiTypography-root-44 MuiTypography-caption-54"
        >
          0
        </span>
      </span>
      <span
        class="MuiTouchRipple-root-82"
      >
        
      </span>
    </button>
    <div
      class="Tweet-flexGrow-5"
    >
      
    </div>
    <button
      aria-label="Go to Tweet"
      class="MuiButtonBase-root-79 MuiIconButton-root-70"
      role="button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label-76"
      >
        <a
          class="Tweet-link-7"
          href="/tweet/test"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-81"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z"
            >
              
            </path>
          </svg>
        </a>
      </span>
      <span
        class="MuiTouchRipple-root-82"
      >
        
      </span>
    </button>
  </div>
</div>
`;
